ledger/participant-integration-api/src/main/scala/platform/store/dao/events/EventsTablePostgresql.scala

```diff
@@ -0,0 +1,231 @@
+          eventIds(i) = eventId.toLedgerString
+          nodeIndexes(i) = nodeId.index
+          flatEventWitnesses(i) = info.stakeholders.getOrElse(nodeId, Set.empty).mkString("|")
+          treeEventWitnesses(i) = info.disclosure.getOrElse(nodeId, Set.empty).mkString("|")
+          createArguments(i) = serialized.createArguments(nodeId)
+          createSignatories(i) = create.signatories.mkString("|")
+          createObservers(i) = create.stakeholders.diff(create.signatories).mkString("|")
+          if (create.coinst.agreementText.nonEmpty) {
+            createAgreementTexts(i) = create.coinst.agreementText
+          }
+          createKeyValues(i) = serialized.createKeyValues.get(nodeId).orNull
+        case exercise: Exercise =>
+          val eventId = EventId(tx.transactionId, nodeId)
+          contractIds(i) = exercise.targetCoid.coid
+          templateIds(i) = exercise.templateId.toString
+          eventIds(i) = eventId.toLedgerString
+          nodeIndexes(i) = nodeId.index
+          flatEventWitnesses(i) = info.stakeholders.getOrElse(nodeId, Set.empty).mkString("|")
+          treeEventWitnesses(i) = info.disclosure.getOrElse(nodeId, Set.empty).mkString("|")
+          exerciseConsuming(i) = exercise.consuming
+          exerciseChoices(i) = exercise.choiceId
+          exerciseArguments(i) = serialized.exerciseArguments(nodeId)
+          exerciseResults(i) = serialized.exerciseResults.get(nodeId).orNull
+          exerciseActors(i) = exercise.actingParties.mkString("|")
+          exerciseChildEventIds(i) = exercise.children
+            .map(EventId(tx.transactionId, _).toLedgerString)
+            .iterator
+            .mkString("|")
+        case _ => throw new UnexpectedNodeException(nodeId, tx.transactionId)
+      }
+    }
+
+    val inserts = insertEvents(
+      eventIds,
+      eventOffsets,
+      contractIds,
+      transactionIds,
+      workflowIds,
+      ledgerEffectiveTimes,
+      templateIds,
+      nodeIndexes,
+      commandIds,
+      applicationIds,
+      submitters,
+      flatEventWitnesses,
+      treeEventWitnesses,
+      createArguments,
+      createSignatories,
+      createObservers,
+      createAgreementTexts,
+      createConsumedAt,
+      createKeyValues,
+      exerciseConsuming,
+      exerciseChoices,
+      exerciseArguments,
+      exerciseResults,
+      exerciseActors,
+      exerciseChildEventIds
+    )
+
+    val archivals =
+      info.archives.iterator.map(archive(tx.offset)).toList
+
+    new Batches(
+      insertEvents = Some(inserts),
+      updateArchives = batch(updateArchived, archivals),
+    )
+  }
+
+  // Specific for PostgreSQL parallel unnesting insertions
+
+  private implicit object BooleanArrayToStatement extends ToStatement[Array[Boolean]] {
+    override def set(s: PreparedStatement, index: Int, v: Array[Boolean]): Unit = {
+      val conn = s.getConnection
+      val bs = conn.createArrayOf("BOOLEAN", v.map(Boolean.box))
```

Comment:
This `map` (and the equivalent in `IntArrayToStatement`) looks expensive. Why not just make the array one of `java.lang.Boolean` to start with?

Reply:
Because Scala optimizes it unless I box explicitly, unfortunately.

Reply:
Are you sure?

```scala
scala> val scalaBools = Array.ofDim[Boolean](5)
scalaBools: Array[Boolean] = Array(false, false, false, false, false)

scala> scalaBools(0)
res5: Boolean = false

scala> scalaBools(0).getClass
res6: Class[Boolean] = boolean

scala> val javaBools = Array.ofDim[java.lang.Boolean](5)
bools: Array[Boolean] = Array(null, null, null, null, null)

scala> javaBools(0)
res3: Boolean = null

scala> javaBools(0).getClass
java.lang.NullPointerException
  ... 28 elided
```

Reply:
Ah, I see you _did_ change it. Cool. ^_^